File: crates/domain/src/models/wrapup.rs
use chrono::NaiveDate;
use uuid::Uuid;

#[derive(Clone, Debug)]
pub struct DateRange {
    pub start: NaiveDate,
    pub end: NaiveDate,
}

#[derive(Clone, Debug)]
pub struct MovieRef {
    pub title: String,
    pub year: u16,
    pub runtime_minutes: Option<u32>,
    pub poster_path: Option<String>,
}

#[derive(Clone, Debug)]
pub struct UserRef {
    pub user_id: Uuid,
    pub display_name: String,
}

#[derive(Clone, Debug)]
pub struct PersonStat {
    pub name: String,
    pub count: u32,
    pub avg_rating: f64,
}

#[derive(Clone, Debug)]
pub struct GenreStat {
    pub genre: String,
    pub count: u32,
    pub avg_rating: f64,
}

#[derive(Clone, Debug)]
pub struct KeywordStat {
    pub keyword: String,
    pub count: u32,
}

#[derive(Clone, Debug)]
pub struct LangStat {
    pub language: String,
    pub count: u32,
}

#[derive(Clone, Debug)]
pub struct MonthCount {
    pub year_month: String,
    pub label: String,
    pub count: u32,
}

#[derive(Clone, Debug)]
pub enum WrapUpScope {
    User(Uuid),
    Global,
}

#[derive(Clone, Debug)]
pub struct WrapUpReport {
    pub scope: WrapUpScope,
    pub date_range: DateRange,
    pub generated_at: chrono::DateTime<chrono::Utc>,

    // Core viewing
    pub total_movies: u32,
    pub total_watch_time_minutes: u32,
    pub movies_per_month: Vec<MonthCount>,
    pub busiest_month: Option<String>,
    pub busiest_day_of_week: Option<String>,
    pub avg_rating: Option<f64>,
    pub rating_distribution: [u32; 5],
    pub longest_movie: Option<MovieRef>,
    pub shortest_movie: Option<MovieRef>,

    // People insights
    pub top_directors: Vec<PersonStat>,
    pub top_actors: Vec<PersonStat>,
    pub director_diversity: u32,
    pub actor_diversity: u32,

    // Genre & taste
    pub top_genres: Vec<GenreStat>,
    pub genre_diversity: u32,
    pub highest_rated_genre: Option<String>,
    pub lowest_rated_genre: Option<String>,
    pub top_keywords: Vec<KeywordStat>,

    // Financial/production (None when data unavailable)
    pub total_budget_watched: Option<i64>,
    pub avg_budget: Option<i64>,
    pub language_distribution: Vec<LangStat>,
    pub oldest_movie: Option<MovieRef>,
    pub newest_movie: Option<MovieRef>,

    // Rewatch stats
    pub total_rewatches: u32,
    pub most_rewatched_movie: Option<MovieRef>,
    pub avg_rating_change_on_rewatch: Option<f64>,

    // Highlights
    pub highest_rated_movie: Option<MovieRef>,
    pub lowest_rated_movie: Option<MovieRef>,
    pub first_movie_of_period: Option<MovieRef>,
    pub last_movie_of_period: Option<MovieRef>,

    // Global-only (None for per-user)
    pub most_active_user: Option<UserRef>,
    pub most_watched_movie_global: Option<MovieRef>,
    pub total_users_active: Option<u32>,

    // Asset references for renderers
    pub poster_paths: Vec<String>,
    pub top_cast_profile_paths: Vec<String>,
}

File: spa/src/features/wrapup.ts
import { z } from "zod"
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query"
import { del, get, post } from "@/lib/api/client"

export const generateWrapUpRequestSchema = z.object({
  start_date: z.string(),
  end_date: z.string(),
  user_id: z.string().uuid().optional(),
})
export type GenerateWrapUpRequest = z.infer<typeof generateWrapUpRequestSchema>

export const wrapUpGeneratedResponseSchema = z.object({
  id: z.string(),
})
export type WrapUpGeneratedResponse = z.infer<typeof wrapUpGeneratedResponseSchema>

export const wrapUpStatusResponseSchema = z.object({
  id: z.string(),
  user_id: z.string().optional(),
  status: z.string(),
  start_date: z.string(),
  end_date: z.string(),
  created_at: z.string(),
  completed_at: z.string().optional(),
  error_message: z.string().optional(),
})
export type WrapUpStatusResponse = z.infer<typeof wrapUpStatusResponseSchema>

export const wrapUpListResponseSchema = z.object({
  items: z.array(wrapUpStatusResponseSchema),
})
export type WrapUpListResponse = z.infer<typeof wrapUpListResponseSchema>

export type MovieRef = {
  movie_id?: string
  title: string
  year: number
  runtime_minutes?: number
  poster_path?: string
}

export type PersonStat = {
  person_id?: string
  name: string
  count: number
  avg_rating: number
}

export type GenreStat = {
  genre: string
  count: number
  avg_rating: number
}

export type MonthCount = {
  year_month: string
  label: string
  count: number
}

export type KeywordStat = {
  keyword: string
  count: number
}

export type LangStat = {
  language: string
  count: number
}

export type WrapUpReport = {
  date_range: { start: string; end: string }
  total_movies: number
  total_watch_time_minutes: number
  movies_per_month: MonthCount[]
  busiest_month?: string
  busiest_day_of_week?: string
  avg_rating?: number
  rating_distribution: number[]
  longest_movie?: MovieRef
  shortest_movie?: MovieRef
  top_directors: PersonStat[]
  top_actors: PersonStat[]
  director_diversity: number
  actor_diversity: number
  top_genres: GenreStat[]
  genre_diversity: number
  highest_rated_genre?: string
  lowest_rated_genre?: string
  top_keywords: KeywordStat[]
  total_budget_watched?: number
  avg_budget?: number
  language_distribution: LangStat[]
  oldest_movie?: MovieRef
  newest_movie?: MovieRef
  total_rewatches: number
  most_rewatched_movie?: MovieRef
  highest_rated_movie?: MovieRef
  lowest_rated_movie?: MovieRef
  first_movie_of_period?: MovieRef
  last_movie_of_period?: MovieRef
  watch_medium_distribution: { medium: string; count: number }[]
  poster_paths: string[]
  top_cast_profile_paths: string[]
}

function generateWrapUp(data: GenerateWrapUpRequest) {
  return post<WrapUpGeneratedResponse>("/wrapups/generate", data)
}

function getWrapUps() {
  return get<WrapUpListResponse>("/wrapups")
}

function getWrapUp(id: string) {
  return get<WrapUpStatusResponse>(`/wrapups/${id}`)
}

function deleteWrapUp(id: string) {
  return del(`/wrapups/${id}`)
}

function getWrapUpReport(id: string) {
  return get<WrapUpReport>(`/wrapups/${id}/report`)
}

export const wrapupKeys = {
  all: ["wrapups"] as const,
  list: () => [...wrapupKeys.all, "list"] as const,
  detail: (id: string) => [...wrapupKeys.all, id] as const,
  report: (id: string) => [...wrapupKeys.all, id, "report"] as const,
}

export function useWrapUpReport(id: string) {
  return useQuery({
    queryKey: wrapupKeys.report(id),
    queryFn: () => getWrapUpReport(id),
    enabled: !!id,
  })
}

export function useWrapUps() {
  return useQuery({
    queryKey: wrapupKeys.list(),
    queryFn: getWrapUps,
  })
}

export function useWrapUp(id: string) {
  return useQuery({
    queryKey: wrapupKeys.detail(id),
    queryFn: () => getWrapUp(id),
    enabled: !!id,
  })
}

export function useGenerateWrapUp() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: GenerateWrapUpRequest) => generateWrapUp(data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: wrapupKeys.all })
    },
  })
}

export function useDeleteWrapUp() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => deleteWrapUp(id),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: wrapupKeys.all })
    },
  })
}
